platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/TaskList.java

```diff
@@ -0,0 +1,215 @@
+   * successful or <code>false</code> if not.
+   *
+   * @param op the operation for which to record the task
+   * @param id a unique identifier for the task (e.g. feature name)
+   * @param task the task to add for later execution
+   * @throws IllegalArgumentException if <code>op</code>, <code>id</code>, or <code>task</code> is
+   *     <code>null</code>
+   */
+  public void add(Operation op, String id, Predicate<ProfileMigrationReport> task) {
+    Validate.notNull(op, "invalid null operation");
+    Validate.notNull(id, "invalid null task id");
+    Validate.notNull(task, "invalid null task");
+    LOGGER.debug("Recording {} task for {} '{}'", op, type, id);
+    report.recordTask();
+    groups
+        .computeIfAbsent(op, o -> new LinkedHashMap<>()) // preserve order of execution
+        .put(id, task);
+  }
+
+  /**
+   * Increases the number of attempts left for the specified operation
+   *
+   * @param op the operation to increase the number of attempts for
+   * @throws IllegalArgumentException if <code>op</code> is <code>null</code>
+   */
+  public void increaseAttemptsFor(Operation op) {
+    Validate.notNull(op, "invalid null operation");
+    // increase the number of attempts left for the given operation and make sure we have at lease 1
+    // left
+    final int n = attemptsLeft.get(op).updateAndGet(a -> (a <= 0) ? 1 : ++a);
+
+    LOGGER.debug("increasing {} task attempts left for {}s to: {}", op, type, n);
+  }
+
+  /**
+   * Checks if this task list is empty.
+   *
+   * @return <code>true</code> if no tasks were added to this task list; <code>false</code> if at
+   *     least one was registered
+   */
+  public boolean isEmpty() {
+    return groups.isEmpty();
+  }
+
+  /**
+   * Gets the the first available operation/group of registered tasks.
+   *
+   * @return the operation associated with the first available group of tasks or empty if no tasks
+   *     were added
+   */
+  public Optional<Operation> getOperation() {
+    return Stream.of(Operation.values()) // search based on defined operation order
+        .filter(groups::containsKey)
+        .findFirst();
+  }
+
+  /**
+   * Executes all tasks defined in the first available operation/group of tasks.
+   *
+   * <p><i>Note:</i> The task list will be cleared unless we have exceeded the maximum number of
+   * attempts for the first available operation.
+   *
+   * @return <code>true</code> if all tasks in the first available operation group were successful;
+   *     <code>false</code> otherwise or if we have exceeded the maximum number of attempts for the
+   *     first available operation
+   * @throws IllegalArgumentException if <code>report</code> is <code>null</code>
+   */
+  public boolean execute() {
+    Validate.notNull(report, "invalid null report");
+    LOGGER.debug("Executing {}s import", type);
+    final Operation op = getOperation().orElse(null);
+
+    if (op == null) { // if no groups have tasks
```

Comment:
❓ Feels odd that we have `getOperation()` return an `Optional` and then we use a `null` check here. Why not use `isPresent()` instead, or just return `null` from `getOperation()`?

Reply:
easy. this is more optimal. If you do `isPresent(), then you will follow it by `get() `. Both methods will internally check if the value is present. Doing one `orElse()` checks only once.